### PR TITLE
New board: AirPi HAT 

### PR DESCRIPTION
**Pull request by SalimTerryLi**
## AirPi HAT for Raspberry Pi B series
It seems the name "AirPi" is already occupied by a project focusing on weather, while I still not finally name this board.

Thanks for every contributor's effort that makes PX4 convenient and powerful as today. This is the PR for a new HAT designed to directly run PX4 on Raspberry Pi B series. It is developed with the inspiration of Navio2, and provides some kinds of different solutions. Maybe better, or worse. I'm not sure.

I'm the author of this board. I started building a vehicle with PX4 running on RPi since 2017, with Dupont Lines and sensor modules at first, and finally come up with this standalone solution recently. The hardware design files locate [here](https://github.com/Airunix-2020-SCU/AirPi2_PCB). On the software side, latest RaspberryPi OS + PX4 will perfectly working. No kernel drivers or proprietary part is needed. Most of driver and modules can be shared with other boards.

### IC selection:
Note that no backup sensors available.
- ICM42605/ICM42688p : currently only 42688p is supported by PX4 but I have some pieces built with 42605.
- IST8310 : as internal mag sensor
- MS5611 : barometer
- PCA9685 : 16 channel PWM output
- ADS1115 : ADC, used for batt voltage monitor, etc
- SC16IS752 : expand UART interface
- NC7SZ86M5 : inverter
- REF5050 : reference voltage for analog sensors

Another part of board provides 5V power supply. It contains:
- XL4015
- L74L06

### Perpheral capability:
- 1x I2C for external compass or others
- 2x UART for GPS and Telemetry
- 1x RC input(actually is UART with only RX), with adjustable hardware inverter
- 16x PWM channels
- 1x battery connector, supports 3~6S
- 4x 16bit ADC channel. One of them is tied to voltage divider that monitors battery voltage
- and, all four USB connectors are available. This board only takes up GPIO resource.

### Issues not resolved:
- [PCA9685](https://github.com/PX4/Firmware/pull/14269)

### Flight logs

https://logs.px4.io/plot_app?log=588e6ff4-f729-4a9a-9828-b69a925ad3f7
https://logs.px4.io/plot_app?log=7f0f21dd-680f-4fd8-85cf-658f7c926878
These two are indoor flight logs, with default PID value, on [this](https://github.com/SalimTerryLi/Firmware/tree/pr-AirPiHAT-newboard-with_drivers) branch. Currently I cannot provide outdoor testing because my campus locates just beside an airport.

### Additional information
Recent status
That is how the first 5 pcs look like. Sensor part(left) and power part(right) will be mounted vertically when the last bit of material arrived.
I'm trying to contact Holybro for some kind of production. Specially thanks for @jinger26 's help.

Any suggestions are welcome, including but not limited to naming, coding, and manufacturing.

### CODE REVIEW

**Comment by bys1123** [comment]
Wow, awesome!

**Comment by dagar** [comment]
Can you give this a new vendor folder (currently `boards/px4`)?

**Comment by SalimTerryLi** [comment]
> 
> 
> Can you give this a new vendor folder (currently `boards/px4`)?

It is just a naming issue. I'd find out a proper name for that. May I ask for the status of #14269 and #14437 ? These two drivers should be the last thing that blocks. Should I wait for them being accepted into upstream? Or close them and merge locally first?

**Comment by SalimTerryLi** [comment]
Flight logs uploaded. Seems that severe vibration occurred? @dagar

**Comment by dagar** [comment]
> Seems that severe vibration occurred?

Where?

**Comment by SalimTerryLi** [comment]
> > Seems that severe vibration occurred?
> 
> Where?

Flight log indicates vibration metrics goes around 0.01? I'm not sure whether it is in acceptable range...

**Comment by SalimTerryLi** [comment]
ADS1115 driver locates inside `drivers/adc`. How can I deal with this naming issue? ADS1115 is a typical ADC so I think it should be placed inside adc directory, but that is already occupied by another kind of driver wrapper. @dagar

File: boards/scumaker/airpi2/src/board_config.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2017 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline boards/scumaker/airpi2/src/board_config.h:3]:
```suggestion
 *   Copyright (c) 2020 PX4 Development Team. All rights reserved.
```

File: src/drivers/drv_sensor.h [hunk 2]
```diff
@@ -163,6 +163,7 @@
 #define DRV_MAG_DEVTYPE_UAVCAN	0x88
 #define DRV_DIST_DEVTYPE_UAVCAN	0x89
 
+#define DRV_ADC_DEVTYPE_ADS1115			0x90
```
Comment (dagar) [inline src/drivers/drv_sensor.h:166]:
```suggestion
#define DRV_ADC_DEVTYPE_ADS1115	0x90
```

**Comment by dagar** [comment]
1. Can you keep `drivers/adc/ADS1115` standalone until we figure out what to do with `drivers/adc` module? Modules within modules seems a bit weird unless they're so tightly coupled they only make sense together.
 2. Add to CI so we don't break anything accidentally.
    - https://github.com/PX4/Firmware/blob/a9798454cdf9c9f2ee5c9e59add5e09c2bc15daa/.ci/Jenkinsfile-compile#L19

**Comment by SalimTerryLi** [comment]
>     1. Can you keep `drivers/adc/ADS1115` standalone until we figure out what to do with `drivers/adc` module? Modules within modules seems a bit weird unless they're so tightly coupled they only make sense together.
> 
>     2. Add to CI so we don't break anything accidentally.
>        
>        * https://github.com/PX4/Firmware/blob/a9798454cdf9c9f2ee5c9e59add5e09c2bc15daa/.ci/Jenkinsfile-compile#L19

I imported ADS1115 branch from #14437 by standalone commit 61c21d9 so a single revert operation can do the job. Should we finish the naming issue in 14437 first?

**Comment by dagar** [comment]
> I imported ADS1115 branch from #14437 by standalone commit [61c21d9](https://github.com/PX4/Firmware/commit/61c21d9d09869fb25a215208ecace9aca9796cf9) so a single revert operation can do the job. Should we finish the naming issue in 14437 first?

Whichever way you'd prefer.

**Comment by SalimTerryLi** [comment]
> Whichever way you'd prefer.

I'd like to finish ADS1115 issue in its PR then revert this branch and do the rebase. Should we move drivers/adc to drivers/adc/buildin_adc, or any other options?

**Comment by SalimTerryLi** [comment]
Please have a look at #14437 and #14477. @dagar It's a series of PRs to solve the issue.

**Comment by SalimTerryLi** [comment]
Any idea about pca9685_pwm_out? @dagar

**Comment by dagar** [comment]
> Any idea about pca9685_pwm_out? @dagar

Anything in particular? I'd still like to centralize this stuff so we can trivially adjust the PWM parameters at runtime and remove the IOCTL/init mess, but this looks fine for now.

**Comment by SalimTerryLi** [comment]
> > Any idea about pca9685_pwm_out? @dagar
> 
> Anything in particular? I'd still like to centralize this stuff so we can trivially adjust the PWM parameters at runtime and remove the IOCTL/init mess, but this looks fine for now.

I cannot make it work with I2CSPIDriver, but it functions normally.

**Review by dagar — APPROVED** [review]

**Pull request merged**